File: app/src/main/java/starbright/com/projectegg/features/ingredients/IngredientsCartAdapter.java
package starbright.com.projectegg.features.ingredients;

import android.content.Context;
import android.support.annotation.NonNull;
import android.support.v7.widget.RecyclerView;
import android.view.LayoutInflater;
import android.view.View;
import android.view.ViewGroup;
import android.widget.ImageView;
import android.widget.TextView;

import com.bumptech.glide.load.engine.DiskCacheStrategy;
import com.bumptech.glide.request.RequestOptions;

import java.util.List;

import butterknife.BindView;
import butterknife.ButterKnife;
import starbright.com.projectegg.R;
import starbright.com.projectegg.data.local.model.Ingredient;
import starbright.com.projectegg.util.GlideApp;

class IngredientsCartAdapter extends RecyclerView.Adapter<IngredientsCartAdapter.ViewHolder> {

    private Context mContext;
    private List<Ingredient> mIngredientCart;
    private Listener mListener;

    IngredientsCartAdapter(Context context, List<Ingredient> cart) {
        mContext = context;
        mIngredientCart = cart;
    }

    @NonNull
    @Override
    public ViewHolder onCreateViewHolder(@NonNull ViewGroup parent, int viewType) {
        View view = LayoutInflater.from(mContext).inflate(R.layout.item_cart, parent, false);
        final ViewHolder viewHolder = new ViewHolder(view);
        viewHolder.ivClear.setOnClickListener(new View.OnClickListener() {
            @Override
            public void onClick(View v) {
                mListener.onClearItemClickedListener(viewHolder.getAdapterPosition());
            }
        });
        return viewHolder;
    }

    @Override
    public void onBindViewHolder(@NonNull ViewHolder holder, int position) {
        holder.bind(mIngredientCart.get(position));
    }

    @Override
    public int getItemCount() {
        return mIngredientCart.size();
    }

    void setListener(Listener listener) {
        mListener = listener;
    }

    class ViewHolder extends RecyclerView.ViewHolder {

        @BindView(R.id.iv_ingredient)
        ImageView ivIngredient;

        @BindView(R.id.tv_ingredient_name)
        TextView tvIngredientName;

        @BindView(R.id.iv_clear)
        ImageView ivClear;

        ViewHolder(View itemView) {
            super(itemView);
            ButterKnife.bind(this, itemView);
        }

        void bind(Ingredient ingredient) {
            tvIngredientName.setText(ingredient.getName());
            GlideApp.with(mContext)
                    .load(ingredient.getImageUrl())
                    .diskCacheStrategy(DiskCacheStrategy.ALL)
                    .apply(RequestOptions.circleCropTransform())
                    .centerCrop()
                    .into(ivIngredient);
        }
    }

    interface Listener {
        void onClearItemClickedListener(int position);
    }
}

File: app/src/main/java/starbright/com/projectegg/features/ingredients/IngredientsCartAdapter.kt
/*
 * Copyright (c) by Andreas (oentoro.andreas@gmail.com)
 * created at 25 - 7 - 2020.
 */

package starbright.com.projectegg.features.ingredients

import android.content.Context
import android.view.LayoutInflater
import android.view.View
import android.view.ViewGroup
import androidx.core.content.ContextCompat
import androidx.recyclerview.widget.RecyclerView
import com.bumptech.glide.load.engine.DiskCacheStrategy
import com.bumptech.glide.request.RequestOptions
import kotlinx.android.synthetic.main.item_cart.view.*
import starbright.com.projectegg.R
import starbright.com.projectegg.data.model.Ingredient
import starbright.com.projectegg.util.GlideApp

class IngredientsCartAdapter(
    private val context: Context,
    private val ingredientCart: List<Ingredient>
) : RecyclerView.Adapter<IngredientsCartAdapter.ViewHolder>() {

    var listener: ((Int) -> (Unit))? = null

    override fun onCreateViewHolder(parent: ViewGroup, viewType: Int): ViewHolder {
        val view = LayoutInflater.from(context).inflate(R.layout.item_cart, parent, false)
        return ViewHolder(view).also { holder ->
            holder.itemView.iv_clear?.setOnClickListener {
                listener?.invoke(holder.adapterPosition)
            }
        }
    }

    override fun onBindViewHolder(holder: ViewHolder, position: Int) =
        holder.bind(ingredientCart[position])

    override fun getItemCount(): Int = ingredientCart.size

    inner class ViewHolder(itemView: View) : RecyclerView.ViewHolder(itemView) {

        fun bind(ingredient: Ingredient) {
            itemView.tv_ingredient_name.text = ingredient.name
            GlideApp.with(context)
                .load(
                    if (ingredient.imageUrl?.isNotEmpty() == true) {
                        ingredient.imageUrl
                    } else {
                        ContextCompat.getDrawable(context, R.drawable.ic_room_service)
                    }
                )
                .diskCacheStrategy(DiskCacheStrategy.ALL)
                .apply(RequestOptions.circleCropTransform())
                .into(itemView.iv_ingredient)
        }
    }
}
